### PR TITLE
Add support for buffer message callbacks

### PR DESCRIPTION
**Pull request by simoncozens**

### CODE REVIEW

File: src/uharfbuzz/_harfbuzz.pyx [hunk 1]
```diff
@@ -21,6 +21,10 @@ cdef extern from "Python.h":
     int PyUnicode_4BYTE_KIND
 
 
+cdef int msgcallback(hb_buffer_t *buffer, hb_font_t *font, const char* message, void* userdata):
+    (<object>userdata)(message)
+    return 1
+
```
Comment (justvanrossum) [inline src/uharfbuzz/_harfbuzz.pyx]: One more blank line here

File: src/uharfbuzz/_harfbuzz.pyx [hunk 2]
```diff
@@ -220,6 +224,8 @@ cdef class Buffer:
     def guess_segment_properties(self) -> None:
         hb_buffer_guess_segment_properties(self._hb_buffer)
 
+    def set_message_func(self, realcallback) -> None:
+        hb_buffer_set_message_func(self._hb_buffer, msgcallback, <void*>realcallback, NULL)
```
Comment (justvanrossum) [inline src/uharfbuzz/_harfbuzz.pyx]: And one more blank line here

File: src/uharfbuzz/charfbuzz.pxd [hunk 3]
```diff
@@ -114,6 +114,13 @@ cdef extern from "hb.h":
     hb_language_t hb_buffer_get_language(hb_buffer_t* buffer)
     void hb_buffer_set_language(hb_buffer_t* buffer, hb_language_t language)
     void hb_buffer_destroy(hb_buffer_t* buffer)
+    ctypedef hb_bool_t   (*hb_buffer_message_func_t) (hb_buffer_t *buffer,
```
Comment (justvanrossum) [inline src/uharfbuzz/charfbuzz.pxd]: Have a look at similar declarations for formatting details, eg. line 150.

File: src/uharfbuzz/charfbuzz.pxd [hunk 4]
```diff
@@ -114,6 +114,13 @@ cdef extern from "hb.h":
...
+                             hb_font_t   *font,
+                             const char  *message,
+                             void        *user_data)
+    void hb_buffer_set_message_func (hb_buffer_t *buffer,
```
Comment (justvanrossum) [inline src/uharfbuzz/charfbuzz.pxd]: Same, have a look at line 85 for example

**Comment by justvanrossum** [comment]
I would have loved to contribute a test, but the included test font has no OTL tables at all.

Maintainers: we could add a simple GSUB feature to the blank subset font. Would that be acceptable?

File: src/uharfbuzz/_harfbuzz.pyx [hunk 5]
```diff
@@ -220,6 +224,8 @@ cdef class Buffer:
     def guess_segment_properties(self) -> None:
         hb_buffer_guess_segment_properties(self._hb_buffer)
 
+    def set_message_func(self, realcallback) -> None:
```
Comment (adrientetar) [inline src/uharfbuzz/_harfbuzz.pyx]: Let's just call this callback

**Comment by justvanrossum** [comment]
Wondering whether the msg string could be passed as an actual string instead of a bytes array?

**Comment by anthrotype** [comment]
@simoncozens thanks for this!

> Maintainers: we could add a simple GSUB feature to the blank subset font. Would that be acceptable?

sgtm

**Comment by justvanrossum** [comment]
I think this PR is good to go. I'll add a test case as a separate PR after the merge.

**Comment by anthrotype** [comment]
@justvanrossum I added you as collaborator with write access, you should be able to merge PRs now.

**Comment by justvanrossum** [comment]
Thanks! I wasn't sure whether I was supposed to, though, as a brand new committer.

**Pull request merged**